### PR TITLE
Added support to override the protocol used when connecting to Oracle.

### PR DESCRIPTION
**Pull request by martinedge**
TCPS is required to be passed in the EasyConnectString to connect to Autonomous databases. Tested against an Oracle Cloud instance using TLS authentication.

|      Q       |   A
|------------- | -----------
| Type         | feature/improvement

Fixes #6593.

#### Summary

Attempted to connect to an autonomous database within Oracle OCI Cloud. Identified the protocol needs to be changed within the connection string to match their requirements. Added a parameter to permit this override. 

TCP remains the default, can be overridden by TCPS and TCPS only.

### CODE REVIEW

File: src/Driver/AbstractOracleDriver/EasyConnectString.php
Comment (morozov) [inline src/Driver/AbstractOracleDriver/EasyConnectString.php]: I believe we should stop putting driver-specific parameters into the driver-agnostic `$params`. Please try moving it to driver options.

Comment (martinedge) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: Hey @morozov - It's an Oracle parameter rather than a driver-specific function. You'll find it used when you generate a connection string from within OCI Cloud.  

It's not dissimilar to how it's done in java - as much as it's driven within an oracle namespace. 

Where else would you suggest I put this if not in an AbstractOracleDriver/EasyConnectString.php ? 

Also noting that 'TCP' as the protocol is already set statically in this portion of code? 

Cheers!

Comment (derrabus) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: > It's an Oracle parameter rather than a driver-specific function.

So… it's a parameter that we only ever need when connecting through the Oracle drivers which makes it very driver-specific.

Comment (martinedge) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: Hi @derrabus - and the change is in AbstractOracleDriver - am I missing something?

Comment (greg0ire) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: As @morozov said, `$params` is driver-agnostic. That's regardless of which class receives it.

Comment (martinedge) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: Ah now I follow what was being said. 

I’ll look at the other drivers to see how specific database settings are made available

Comment (martinedge) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: @greg0ire @derrabus @morozov I have moved the configuration item into driverOptions, in line with the 'exclusive' configuration item that already existed. Hopefully, that's better.

File: src/Driver/AbstractOracleDriver/EasyConnectString.php
Comment (morozov) [inline src/Driver/AbstractOracleDriver/EasyConnectString.php]: Why hard-code the allowed values? The logic here might be as simple as:
```php
$protocol = $params['driverOptions']['protocol'] ?? 'TCP';
```

File: tests/Driver/OCI8/DriverTest.php
Comment (morozov) [inline tests/Driver/OCI8/DriverTest.php]: It would be better to have a functional test instead. Is it possible to query some system view and get the protocol of the current connection? Otherwise, this test doesn't prove anything.

File: docs/en/reference/configuration.rst
Comment (morozov) [inline docs/en/reference/configuration.rst]: It looks like Autonomous is an Oracle product. Is it relevant in this documentation? An explanation of what TCPS is or a link to documentation would make more sense.

File: src/Driver/AbstractOracleDriver/EasyConnectString.php
Comment (martinedge) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: @morozov I guess I was trying to restrict the chance of someone using the wrong thing. If that's not something for me to worry about, I can do that.

File: tests/Driver/OCI8/DriverTest.php
Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov I can check, but if you got the protocol wrong, it would not connect in the first place, so would I not be creating a test that would in theory never succeed?

File: src/Driver/AbstractOracleDriver/EasyConnectString.php
Comment (morozov) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: It's the job of the database or the driver to worry about that. Otherwise, DBAL may forbid the usage of potentially valid values.

File: tests/Driver/OCI8/DriverTest.php
Comment (morozov) [reply tests/Driver/OCI8/DriverTest.php]: I'm talking about testing the happy path. If the user provides a valid value, the connection should be established using this protocol.

File: src/Driver/AbstractOracleDriver/EasyConnectString.php
Comment (martinedge) [reply src/Driver/AbstractOracleDriver/EasyConnectString.php]: I have updated the code to use your suggestion, but I moved the default into the connect string directly

File: tests/Driver/OCI8/DriverTest.php
Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov OK. looking into it.

File: docs/en/reference/configuration.rst
Comment (martinedge) [reply docs/en/reference/configuration.rst]: I have updated it with a link to TCPS on Oracle.

File: tests/Driver/OCI8/DriverTest.php
Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov The closest I can find in Oracle is via v$session_conn_info:

select * from V$SESSION_CONNECT_INFO;

` (DESCRIPTION=(address=(protocol=tcps)(port=1521)(host=1.1.1.1))(CONNECT_DATA=(CID=(PROGRAM=DataGrip)(HOST=__jdbc__)(USER=xxxx))(service_name=xxxxx.oraclecloud.com)(CONNECTION_ID=xxxxxxx)(SERVER=dedicated)(INSTANCE_NAME=xxxx))(ADDRESS=(PROTOCOL=TCP)(HOST=1.1.1.1)(PORT=1521)(publish_service_acl=true))(SOURCE_ROUTE=YES)(HOP_COUNT=2))
`
I would have to parse that string then to get the answer unless a pattern match would be OK. I'm working on the 'existence' of protocol=tcps - because there may be some magic in Oracle cloud where it is likely passing via TCP internally.

What are your thoughts?

Comment (morozov) [reply tests/Driver/OCI8/DriverTest.php]: I think it's okay. If Oracle reports this parameter back, it means that it's passed to Oracle and it's valid.

> I'm working on the 'existence' of protocol=tcps

This is fine. I would structure the test this way:
1. Establish a connection w/ the default protocol. Check that the connection sting doesn't contain this substring.
2. Establish a connection w/ the other protocol. Check that the connection sting contains this substring.

This way, we will make sure that the substring appears in the connection string as a result of configuration change, not on its own. Also, those will be two independent tests (or one with a data provider).

Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov i will look shortly

Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov I've given it a shot, there is an issue though

* If it works via TCPS, chances are it won't connect using TCP
* Earlier versions (sub version 21) don't contain the connect string column in the view. 

This works in my 'successful' test case, on the right database version

`$parameters            = ['envContext' => 'userEnv', 'sidContext' => 'sid', 'protocolString' => '%protocol=tcps%', 'networkServiceBanner' => '%tcp/ip%'];`

`$tcpProtocolCheckQuery = 'with dbdetails as (select sys_context(:envContext, :sidContext) as sid from dual) SELECT 1 FROM v$session_connect_info sci join dbdetails on dbdetails.sid = sci.SID where lower(sci.client_connection_string) like :protocolString and lower(sci.network_service_banner) like :networkServiceBanner';`

Comment (morozov) [reply tests/Driver/OCI8/DriverTest.php]: Got it. In this case, instead of adding a test here, please update this test: https://github.com/doctrine/dbal/blob/5a9b576741cba40dfd9f63af0ead729ad825bdeb/tests/Driver/AbstractOracleDriver/EasyConnectStringTest.php#L49-L59

Add the new parameter to the input and update the expected value accordingly.

Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: Hey @morozov  (Thanks for your help by the way) - I have updated the tests accordingly and tested the test.

Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov oh - and I created a separate test rather than extend that test, is that ok?

File: docs/en/reference/configuration.rst
Comment (derrabus) [inline docs/en/reference/configuration.rst]: Please remove the extra line you've introduced.

Comment (martinedge) [reply docs/en/reference/configuration.rst]: Done

File: tests/Driver/OCI8/DriverTest.php
Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov Just checking if there is anything else you want me to do here  - or we're all good?

Comment (morozov) [reply tests/Driver/OCI8/DriverTest.php]: Yeah, this is fine too. Please squash all commits, and we can merge this.

Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov First time 'squashing' - I do hope I got it right.

Comment (greg0ire) [reply tests/Driver/OCI8/DriverTest.php]: I don't think so, but here is a guide:



1. `git rebase -i origin/4.2.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already pushed. Don't forget the `--force` option otherwise git will try to merge both things together.

Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @greg0ire I believe I have it this time - although the message is a bit skewiff with the note about 'combination of 2 commits'. Says 1 commit now :)

Comment (greg0ire) [reply tests/Driver/OCI8/DriverTest.php]: Regarding the number of commits, we're good now, congrats!

Regarding the message itself, I don't know how you managed to do that :rofl: 

Can tell me what `git config --get 'commit.cleanup'` prints? I suspect an unexpected configuration on your end.

If I run `git commit --amend`, the comments disappear, as expected.
Can you try `git commit --amend --no-edit --cleanup=strip && git push --force`?

Comment (morozov) [reply tests/Driver/OCI8/DriverTest.php]: @greg0ire, thanks for guiding @martinedge here. If it takes too much effort on your end, and if the author doesn't mind, we could amend the commit ourselves.

Comment (martinedge) [reply tests/Driver/OCI8/DriverTest.php]: @morozov I don't mind, if you want me to work through the process, I'm happy to either way, as it might save me mucking it up in the future :-) Up to you all, thankyou for your support!

**Review by morozov — APPROVED** [review]

File: tests/Driver/OCI8/DriverTest.php
Comment (morozov) [reply tests/Driver/OCI8/DriverTest.php]: I did the regular `git commit --amend && git push --force`.

**Comment by morozov** [comment]
Thanks, @martinedge.

**Comment by morozov** [comment]
Actually, it should have gone to 4.3.x since it's a new feature but it looks safe. Maintainers, please let me know if you think we should revert and retarget.

**Comment by greg0ire** [comment]
It looks safe indeed. I think one way to notice that kind of mistake is to form the habit of assigning the milestone right before merging.

**Pull request merged**
